src/rdb.c

```diff
@@ -2185,6 +2217,162 @@ robj *rdbLoadObject(int rdbtype, rio *rdb, sds key, int dbid, int *error) {
+                        decrRefCount(o);
+                        sdsfree(field_dup);
+                        sdsfree(value);
+                        sdsfree(field);
+                        return NULL;
+                    }
+                }
+
+                /* check if the values can be saved to listpack (or should convert to dict encoding) */
+                if (sdslen(field) > server.hash_max_listpack_value ||
+                    sdslen(value) > server.hash_max_listpack_value ||
+                    !lpSafeToAdd(((listpackEx*)o->ptr)->lp, sdslen(field) + sdslen(value) + lpEntrySizeInteger(expire)))
+                {
+                    /* convert to hash */
+                    hashTypeConvert(o, OBJ_ENCODING_HT, hexpires);
+
+                    if (len > DICT_HT_INITIAL_SIZE) { /* TODO: this is NOT the original len, but this is also the case for simple hash, is this a bug? */
+                        if (dictTryExpand(o->ptr, len) != DICT_OK) {
+                            rdbReportCorruptRDB("OOM in dictTryExpand %llu", (unsigned long long)len);
+                            decrRefCount(o);
+                            if (dupSearchDict != NULL) dictRelease(dupSearchDict);
+                            sdsfree(value);
+                            sdsfree(field);
+                            return NULL;
+                        }
+                    }
+
+                    /* don't add the values to the new hash: the next if will catch and the values will be added there */
+                } else {
+                    listpackExAddNew(o, field, value, expire);
+                    sdsfree(field);
+                    sdsfree(value);
+                }
+            }
+
+            if (o->encoding == OBJ_ENCODING_HT) {
+                /* WA for check-rdb mode, when there's no DB so can't attach expired items to ebuckets,
+                 * or when no expiry was not set for this field */
+                if ((db == NULL) || (expire == 0)) {
+                    hashTypeSet(db, o, field, value, 0);
+                } else {
+                    if (hashTypeSetExRdb(db, o, field, value, expire) != C_OK) {
+                        serverLog(LL_WARNING, "failed adding hash field %s to key %s", field, key);
+                        decrRefCount(o);
+                        if (dupSearchDict != NULL) dictRelease(dupSearchDict);
+                        sdsfree(value);
+                        sdsfree(field);
+                        return NULL;
+                    }
+                }
+                sdsfree(field);
+                sdsfree(value);
+            }
+        }
+
+        if (dupSearchDict != NULL) dictRelease(dupSearchDict);
+
+        /* check for empty key (if all fields were expired) */
+        if (hashTypeLength(o, 0) == 0) {
+            decrRefCount(o);
+            goto emptykey;
+        }
+        if ((db != NULL) && (minExpire != EB_EXPIRE_TIME_INVALID))
```

Comment:
Following this change, it won't convert any more to HT in case of `db` is NULL. 

Instead I think we should modify `hashTypeConvert()` to accept pointer of `redisDb` instead of `ebuckets hexpires` and let it handle with NULL. It will simplifes conditions at rdb.c as well.

Reply:
Not sure I follow - what does this change have to do with conversion to dict? Any why should we care about conversion to dict when db is NULL? In this case, the requirement (AFAIK) is to check RDB validity, the resulting object should be discarded (at least in this repo). What am I missing?

Reply:
You don't miss anything. Just saw a change in the logic which i tend to avoid if not necessary at large changes. And in addition to the fact that passing `redisDb` to `hashTypeConvert()` it will relax the condtions at `rdb.c`, made me to write my thought.